### PR TITLE
SRX-63BV32 If config.json file is invalid, give a fatal error during repository initialization

### PR DESCRIPTION
**Pull request by mnishamk-freiheit**
https://revolution.dev/app/-JqFGExX46gs9mH7vxR5/-M37cZx7XYoOvQ5Mmcir/FILESYSTEM/FOLDER/ROOT/STORY/-Mp6i5lzSJREYmTNHy6p

When the config.json file in any of the environments is an invalid json file (missing file is ok) then do not start kuberpult cd-service, throw a valid error message and quit, do the same in /health check as well.

### CODE REVIEW

File: services/cd-service/pkg/service/service.go [hunk 1]
```diff
@@ -72,8 +72,13 @@ func (s *Service) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 }
 
 func (s *Service) ServeHTTPHealth(w http.ResponseWriter, r *http.Request) {
-	w.WriteHeader(200)
-	fmt.Fprintf(w, "ok\n")
+	if err := s.Repository.State().ValidateConfigs(); err != nil {
```
Comment (mnishamk-freiheit) [inline services/cd-service/pkg/service/service.go:75]: This seems a bit expensive to do in every health check

**Comment by mouhsen-ibrahim** [comment]
@mnishamk-freiheit please add tests

File: services/cd-service/pkg/repository/repository.go [hunk 2]
```diff
@@ -891,6 +912,18 @@ func names(fs billy.Filesystem, path string) ([]string, error) {
 	return result, nil
 }
 
+func validateJsonFile(fs billy.Filesystem, path string) (error) {
+	if buf, err := readFile(fs, path); err != nil {
+		return wrapFileError(err, path, "could not read json file")
+	} else {
+		valid := json.Valid(buf)
+		if (!valid) {
+			return errors.New("JSON file " + path + " is invalid.")
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/repository.go:921]:
```suggestion
			return errors.New("JSON file " + path + " is not valid json.")
```

File: services/cd-service/pkg/service/service.go (hunk 1 above)
Comment (sven-urbanski-freiheit-com) [reply services/cd-service/pkg/service/service.go:75]: I agree. Let's remove this from the health check.

File: services/cd-service/pkg/repository/repository.go [hunk 3]
```diff
@@ -208,7 +224,12 @@ func New(ctx context.Context, cfg Config) (Repository, error) {
 				}
 			}
 			// check that we can build the current state
-			if _, err := result.buildState(); err != nil {
+			state, err := result.buildState();
+			if err != nil {
+				return nil, err
+			}
+			
+			if err := state.validateConfigs(); err != nil {
```
Comment (hannesg) [inline services/cd-service/pkg/repository/repository.go:232]: you could just use `state` and check whether `GetEnvironmentConfigs` returns an error. Then you would also check that the configs are semantically valid.

Comment (mnishamk-freiheit) [reply services/cd-service/pkg/repository/repository.go:232]: Had originally discussed this with @sven-urbanski-freiheit-com , and he mentioned to keep validity check without semantics which is why the valid function was added

Comment (sven-urbanski-freiheit-com) [reply services/cd-service/pkg/repository/repository.go:232]: `semantically valid` is a bit of a strong word for the checks we do there ;) But yes, we check a bit more.

But yes, I would agree with Hannes, you can basically just call `GetEnvironmentConfigs` and get a better result.
Just make sure that the error returned is actually readible (but that should already be the case) @mnishamk-freiheit

Comment (mnishamk-freiheit) [reply services/cd-service/pkg/repository/repository.go:232]: ,"msg":"repository.new.error","error":"invalid character ':' after object key:value pair",

this is the error if you use GetEnvironmnetConfigs, it comes from json.decoder.decode method. Might want to wrap this error in something readable pointing to the actual file

File: services/cd-service/pkg/repository/repository_test.go [hunk 4]
```diff
@@ -246,6 +246,141 @@ func TestNew(t *testing.T) {
 					t.Errorf("mismatched file content, expected %s, got %s", "hello", content)
 				}
 			},
+		}, {
+			Name:   "Old with valid config initializes correctly",
+			Branch: "master",
+			Setup:  func(t *testing.T, remoteDir, localDir string) {
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/repository_test.go:252]: please make the 2 new tests into a seperate table driven test, as discussed.

**Pull request merged**